internal/graph/graph.go

```diff
@@ -52,120 +50,101 @@ type ResolutionMetadata struct {
 
-	// NoFurtherEvalCondition indicates an ingress condition whereby results are factual and
+	// NoFurtherEvalCondition indicates an edge condition whereby results found with ReverseExpandQuery are factual and
 	// known to be true and require no further evaluation before a determination of the outcome
 	// can be made.
 	NoFurtherEvalCondition
 )
 
-// RelationshipIngress represents a possible ingress point between some source object reference
-// and a target user reference.
-type RelationshipIngress struct {
+// RelationshipEdge represents a possible relationship some source object reference
+// and a target user reference. The possibility is realized depending on the tuples and on the edge's type.
+type RelationshipEdge struct {
+	Type RelationshipEdgeType
 
-	// The type of the relationship ingress
-	Type RelationshipIngressType
+	// The edge is directed towards this node, which can be like group:*, or group, or group:member
+	TailUsersetNode *openfgav1.RelationReference
 
-	// The relationship reference that defines the ingress to some target relation
-	Ingress *openfgav1.RelationReference
-
-	// TuplesetRelation defines the tupleset relation reference that relates a source
-	// object reference with a target if the type of the relationship ingress is that
-	// of a TupleToUserset
-	// TODO this can be just a string for the relation (since the type will be the same as Ingress.Type)
+	// If the type is TupleToUsersetEdge, this defines the TTU condition
+	// TODO this can be just a string for the relation (since the type will be the same as TailUsersetNode.Type)
 	TuplesetRelation *openfgav1.RelationReference
 
-	Condition IngressCondition
+	// TODO this is leaking implementation details of ReverseExpand. This can be a boolean saying
+	// if `TailUsersetNode` is intersection or exclusion.
+	Condition EdgeCondition
 }
 
-func (r RelationshipIngress) String() string {
+func (r RelationshipEdge) String() string {
 	// TODO also print the condition
 	val := ""
 	if r.TuplesetRelation != nil {
-		val = fmt.Sprintf("ingress %s, type %s, tupleset %s", r.Ingress.String(), r.Type.String(), r.TuplesetRelation.String())
+		val = fmt.Sprintf("userset %s, type %s, tupleset %s", r.TailUsersetNode.String(), r.Type.String(), r.TuplesetRelation.String())
 	} else {
-		val = fmt.Sprintf("ingress %s, type %s", r.Ingress.String(), r.Type.String())
+		val = fmt.Sprintf("userset %s, type %s", r.TailUsersetNode.String(), r.Type.String())
 	}
 	return strings.ReplaceAll(val, "  ", " ")
 }
 
-// ConnectedObjectGraph represents a graph of relationships and the connectivity between
+// TypesystemGraph represents a graph of relationships and the connectivity between
 // object and relation references within the graph through direct or indirect relationships.
-// The ConnectedObjectGraph should be used to introspect what kind of relationships between
-// object types can exist.
-type ConnectedObjectGraph struct {
+type TypesystemGraph struct {
```

Comment:
Can we keep the common theme here and name this `RelationshipGraph`.

We commonly refer to the graph as the "graph of relationships", so it seems appropriate to call this thing the `RelationshipGraph`.